File: core/pie.taskdefs.guice/src/main/java/mb/pie/taskdefs/guice/GuiceTaskDefs.java
package mb.pie.taskdefs.guice;

import com.google.inject.Inject;
import mb.pie.api.MapTaskDefs;
import mb.pie.api.TaskDef;
import mb.pie.api.TaskDefs;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.Map;

/**
 * Task definitions implementation that injects task definition binding created from {@link TaskDefsModule}.
 */
public class GuiceTaskDefs implements TaskDefs {
    private final MapTaskDefs taskDefs;

    /**
     * @throws IllegalArgumentException when there is a duplicate {@link TaskDef#getId() task definition identifier}.
     */
    @Inject public GuiceTaskDefs(Map<String, TaskDef<?, ?>> taskDefs) {
        this.taskDefs = new MapTaskDefs(taskDefs);
    }

    @Override public @Nullable TaskDef<?, ?> getTaskDef(String id) {
        return taskDefs.getTaskDef(id);
    }

    @Override public Iterable<TaskDef<?, ?>> getAllTaskDefs() {
        return taskDefs.getAllTaskDefs();
    }

    @Override public boolean exists(String id) {
        return taskDefs.exists(id);
    }
}

File: taskdefs.guice/src/main/kotlin/mb/pie/taskdefs/guice/GuiceTaskDefs.kt
package mb.pie.taskdefs.guice

import com.google.inject.*
import com.google.inject.multibindings.MapBinder
import mb.pie.api.*
import mb.pie.runtime.taskdefs.MapTaskDefs

/**
 * Sets the task definitions of this builder to the [GuiceTaskDefs] retrieved from given [injector].
 */
fun PieBuilder.withGuiceTaskDefs(injector: Injector): PieBuilder {
  val taskDefs = injector.getInstance(GuiceTaskDefs::class.java)
  this.withTaskDefs(taskDefs)
  return this
}

/**
 * Sets the task definitions of this builder to the given [taskDefs].
 */
fun PieBuilder.withGuiceTaskDefs(taskDefs: GuiceTaskDefs): PieBuilder {
  this.withTaskDefs(taskDefs)
  return this
}


/**
 * [TaskDefs] implementation that injects the map binding created from [TaskDefsModule].
 */
class GuiceTaskDefs @Inject constructor(map: HashMap<String, TaskDef<*, *>>) : MapTaskDefs(map)

/**
 * A module that binds [GuiceTaskDefs] as a singleton, which can be passed to a [PieBuilder] with [withGuiceTaskDefs].
 */
class GuiceTaskDefsModule : Module {
  override fun configure(binder: Binder) {
    binder.bind(GuiceTaskDefs::class.java).`in`(Singleton::class.java)
  }
}


/**
 * A module for binding task definitions. Extend this module, and implement the [bindTaskDefs] function by making calls to [bindTaskDef] to
 * bind task definitions.
 */
abstract class TaskDefsModule : Module {
  override fun configure(binder: Binder) {
    val taskDefsBinder = MapBinder.newMapBinder<String, TaskDef<*, *>>(binder, object : TypeLiteral<String>() {}, object : TypeLiteral<TaskDef<*, *>>() {})
    binder.bindTaskDefs(taskDefsBinder)
  }

  abstract fun Binder.bindTaskDefs(taskDefsBinder: MapBinder<String, TaskDef<*, *>>)

  protected inline fun <reified B : TaskDef<*, *>> Binder.bindTaskDef(builderBinder: MapBinder<String, TaskDef<*, *>>, id: String) {
    bind(B::class.java).`in`(Singleton::class.java)
    builderBinder.addBinding(id).to(B::class.java)
  }
}
